src/runtime-rs/src/shim/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+    #[error("failed to set logger")]
+    Logger,
+    #[error("error from config")]
+    Config(#[source] config::ConfigError),
+    #[error("error from sandbox")]
+    Sandbox(#[source] virtcontainers::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+const SOCKET_ROOT: &str = "/run/containerd";
+
+const ENV_SOCKET_FD: &str = "BIND_FD";
+
+/// Received commandline arguments or environment arguments.
+///
+/// For defail information, please refer to the
+/// [shim v2 spec](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md).
+#[derive(Debug, Default, Clone)]
+pub struct ShimArgs {
+    /// the id of the container
+    pub id: String,
+    /// the namespace for the container
+    pub namespace: String,
+    /// the address of the containerd's main socket
+    pub address: String,
+    /// the binary path to publish events back to containerd
+    pub publish_binary: String,
+    /// the path to the bundle to delete
+    pub bundle: String,
+    /// Whether or not to enable debug
+    pub debug: bool,
+}
+
+impl ShimArgs {
+    /// Check the shim argument object is vaild or not.
+    ///
+    /// The id, namespace, address and publish_binary are mandatory for START, RUN and DELETE.
+    /// And bundle is mandatory for DELETE.
+    pub fn validate(&mut self, is_delete: bool) -> Result<()> {
+        if self.id.is_empty()
+            || self.namespace.is_empty()
+            || self.address.is_empty()
+            || self.publish_binary.is_empty()
+        {
+            return Err(Error::InvalidArgument);
+        }
+        if is_delete && self.bundle.is_empty() {
+            return Err(Error::InvalidArgument);
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(Error::InvalidArgument);
+        }
+
+        // Ensure `address` is a valid path.
+        let path = PathBuf::from(self.address.clone())
+            .canonicalize()
+            .map_err(|_| Error::InvalidArgument)?;
+        let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+        if !md.file_type().is_socket() {
+            return Err(Error::InvalidArgument);
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
+            let path = PathBuf::from(self.bundle.clone())
+                .canonicalize()
+                .map_err(|_| Error::InvalidArgument)?;
+            let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+            if !md.is_dir() {
+                return Err(Error::InvalidArgument);
+            }
+            self.bundle = path
+                .to_str()
+                .map(|v| v.to_owned())
+                .ok_or(Error::InvalidArgument)?;
+        }
+
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
```

Comment:
I like this sort of check, but what about:

- If `comp` is `""`?
- If the path contains unicode values (including the various unicode whitespace values)?
- if the overall path container `../` or `../../../` *etc*?

I'd consider inverting the logic to only accepting printable characters and non-whitespace characters, but by using the standard facilities like https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace and we're bound to miss lots of scenarios.

Reply:
Fixed in the first revision. Use more strict rules: only allow ascii alphanumeric character and '-' and '_'

Reply:
Fixed in the second revision: now allow ascii alphanumeric character and '-', '_', '.' and '~'